fix: Delay the cast from IAmazonDynamoDB to AmazonDynamoDBClient in the DocumentModel

ashovlin opened the pull request:
## Description

**Before:** When initializing the document model `Table` object with an `IAmazonDynamoDB`, .NET/Standard/Core, we cast it right away to an `AmazonDynamoDBClient`. 
* Because we didn't have an synchronous HTTP client, we removed the public sync apis from the low-level service client.
* However we never removed the synchronous public `Table` APIs, which still rely on the internal sync low-level APIs internally.
* So to work around this, we kept the sync low-level APIs as `internal` on `AmazonDynamoDBClient`. This is why we need the cast, since the sync low-level APIs are no longer part of the public `IAmazonDynamoDB`.

**Now:** We delay the cast until right before we need to access the internal, low-level sync APIs.
* For users still calling the sync APIs in .NET/Standard/Core, you can avoid this by moving to the async equivalents where possible.
* If you're still relying on the sync APIs but aren't providing an actual `AmazonDynamoDBClient`, you'll now see a clearer `InvalidOperationException` instead of the `NullReferenceException` from the failed cast.

**Future:**
1. We can deprecate the `Table` sync-over-async APIs entirely. This would remove the need for this cast.
2. Since .NET 5 added back a sync HTTP client, we've discussed internally re-exposing sync low-level APIs where possible.


## Motivation and Context
DOTNET-6755 and #1589 (most of the way, until we do on of the _Future_ ideas from above)

This allows `Table` consumers to mock the `IAmazonDynamoDB` from .NET/Standard/Core as long as you stick to async public APIs.

## Testing
```
 dotnet test .\sdk\src\Services\DynamoDBv2\DynamoDBv2.sln
dotnet test .\sdk\AWSSDK.CoreAndCustomUnitTests.NetStandard.sln
```
PENDING - .NET Standard integ tests for DDB

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project
- [X] My change requires a change to the documentation
- [X] I have updated the documentation accordingly
- [X] I have read the **README** document
- [X] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License
<!--- The SDK is released under the [Apache 2.0 license][license], so any code you submit will be released under that license -->
<!--- For substantial contributions, we may ask you to sign a [Contributor License Agreement (CLA)][cla] -->
<!--- Put an `x` in the below box if you confirm that this request can be released under the Apache 2 license -->
- [X] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

normj commented on sdk/src/Services/DynamoDBv2/Custom/DocumentModel/DocumentBatchGet.cs:
```diff
@@ -427,15 +427,22 @@ private Results GetAttributeItems()
-                var serviceResponse = client.BatchGetItem(request);
+#if AWS_ASYNC_API
+                // Cast the IAmazonDynamoDB to the concrete client instead, so we can access the internal sync-over-async methods
+                var internalClient = client as AmazonDynamoDBClient;
```
nit: You can move the casting logic outside of the loop and only do it once.

ashovlin replied:
Fixed in https://github.com/aws/aws-sdk-net/pull/3388/commits/aa1e3273fae0ba8826c15a91902728713e06a1d4

normj commented on sdk/src/Services/DynamoDBv2/Custom/DocumentModel/DocumentBatchWrite.cs:
```diff
@@ -469,15 +469,22 @@ private static int GetNumberOfWrites(BatchWriteItemRequest request)
             {
-                var result = client.BatchWriteItem(request);
+#if AWS_ASYNC_API
+                // Cast the IAmazonDynamoDB to the concrete client instead, so we can access the internal sync-over-async methods
```
nit: Move the casting logic above the loop.

ashovlin replied:
Fixed in https://github.com/aws/aws-sdk-net/pull/3388/commits/aa1e3273fae0ba8826c15a91902728713e06a1d4

normj reviewed (approved):
A couple minor nits but change looks good.

ashovlin commented:
Realized that I was using too wide of a condition `AWS_ASYNC_API`, which was casting unnecessarily for .NET Framework 4.7.2 even though it still has sync APIs we can use. Switched to `NETSTANDARD` in https://github.com/aws/aws-sdk-net/pull/3388/commits/aa1e3273fae0ba8826c15a91902728713e06a1d4

This also made some .NET Framework tests invalid, so I removed them.

ashovlin commented:
Dry-run e8acaff2-7604-4ade-a913-8ea4cc934cfe passed after adding https://github.com/aws/aws-sdk-net/pull/3388/commits/2f8a80ea5900768e8fe24d18d513fed4f61a085e
